### PR TITLE
Theano shape inference

### PR DESCRIPTION
**Pull request by abhaikollara**
Added shape inference in Theano for
- `flatten`
- `batch_flatten`
- `repeat_elements`
- `squeeze`
- `permute_dimensions`
- `repeat`
- `expand_dims`
- `spatial_2d_padding`
- `tile`

### CODE REVIEW

File: keras/backend/theano_backend.py [hunk 1]
```diff
@@ -682,8 +684,12 @@ def repeat_elements(x, rep, axis):
     If x has shape (s1, s2, s3) and axis=1, the output
     will have shape (s1, s2 * rep, s3).
     """
-    # TODO: `keras_shape` inference.
-    return T.repeat(x, rep, axis=axis)
+    y = T.repeat(x, rep, axis=axis)
+    if hasattr(x, '_keras_shape'):
+        y._keras_shape = list(x._keras_shape)
+        y._keras_shape[axis] = x._keras_shape[axis]*rep
```
Comment (fchollet) [inline keras/backend/theano_backend.py]: Space around operators

Comment (abhaikollara) [inline keras/backend/theano_backend.py]: Fixed it

**Comment by abhaikollara** [comment]
Shape inference for  `gather`,`transpose`,`dot`,`batched_dot` are to be completed too. Shall I complete those before moving on to writing the tests

**Comment by abhaikollara** [comment]
Shape inference for `tile` still has some issues. I'm closing this PR and opening a new one

**Pull request closed without merging**